src/cmd/cmd.zig

```diff
@@ -643,30 +672,34 @@ fn getEntrypoints(logger: Logger) !std.ArrayList(SocketAddr) {
             // use first A record address
             const ipv4_addr = addr_list.addrs[0];
 
-            // parse port from string
-            const port = std.fmt.parseInt(u16, port_str, 10) catch {
-                logger.field("entrypoint", entrypoint).err("entrypoint port not valid");
-                return error.EntrypointPortNotValid;
-            };
-
-            var socket_addr = SocketAddr.fromIpV4Address(ipv4_addr);
-            socket_addr.setPort(port);
+            const socket_addr = SocketAddr.fromIpV4Address(ipv4_addr);
+            std.debug.assert(socket_addr.port() == port);
             break :brk socket_addr;
         };
 
-        try entrypoints.append(socket_addr);
+        const gop = entrypoint_set.getOrPutAssumeCapacityAdapted(socket_addr, EntrypointCtx{ .entrypoints = entrypoints.items });
+        if (!gop.found_existing) {
+            entrypoints.appendAssumeCapacity(socket_addr);
+        }
     }
 
     // log entrypoints
-    var entrypoint_string = try gpa_allocator.alloc(u8, 53 * entrypoints.items.len);
-    defer gpa_allocator.free(entrypoint_string);
-    var stream = std.io.fixedBufferStream(entrypoint_string);
-    var writer = stream.writer();
-    for (0.., entrypoints.items) |i, entrypoint| {
-        try entrypoint.toAddress().format("", .{}, writer);
-        if (i != entrypoints.items.len - 1) try writer.writeAll(", ");
-    }
-    logger.infof("entrypoints: {s}", .{entrypoint_string[0..stream.pos]});
+    const EntrypointsFmt = struct {
+        entrypoints: []const SocketAddr,
+
+        pub fn format(
+            entrypoints_fmt: @This(),
+            comptime fmt_str: []const u8,
+            fmt_options: std.fmt.FormatOptions,
+            writer: anytype,
+        ) !void {
+            for (0.., entrypoints_fmt.entrypoints) |i, entrypoint| {
+                if (i != 0) try writer.writeAll(", ");
+                try entrypoint.toAddress().format(fmt_str, fmt_options, writer);
+            }
+        }
+    };
```

Comment:
can you pull this struct out of the function? struct definitions inside isnt functions isnt standard and we cant reuse the logic elsewhere

Reply:
Is there somewhere you'd want to re-use this? Unscoping this single-use formatter doesn't seem terribly useful compared to just defining it exactly where it's needed.

Also, to be clear, by "standard", do you just mean according to the sig style guide?

Reply:
f761151

Reply:
does this approach still call toAddress on the entrypoints before formatting?

Reply:
It doesn't, but it also doesn't seem to change the behavior, because the `format` method is implemented directly on `SocketAddr`. Is there some substantially different behavior in some edge case?

Reply:
yeah great point there is no difference - can you remove all the formatting code and just log entrypoints.items directly

Reply:
well, that would make it different, the default format for a list is `{ item1, item2 }`

Reply:
that's ok the main goal is seeing the ip and ports - a extra {} is ok

Reply:
f8e0fbddaa2ca457bf1bbda52607a144072707b9